bellman/src/domain/fft.rs

```diff
@@ -0,0 +1,177 @@
+use crate::{multicore, arith};
+use multicore::Worker;
+use arith::Group;
+
+use ff::{Field, ScalarEngine};
+
+pub fn run_optimal_fft<E,T>(a: &mut [T], worker: &Worker, omega: &E::Fr, log_n: u32) 
+where
+    E: ScalarEngine,
+    for <'a> T: Group<'a,E> 
+{
+    let log_cpus = worker.log_num_cpus();
+
+    if log_n <= log_cpus {
+        serial_fft(a, omega, log_n);
+    } else {
+        parallel_fft(a, worker, omega, log_n, log_cpus);
+    }
+}
+
+pub fn serial_fft<E,T>(series: &mut [T], omega: &E::Fr, log_n: u32) 
+where 
+    E: ScalarEngine, 
+    for <'a> T: Group<'a,E>
+{
+    let len: u32 = series.len() as u32;
+    assert_eq!(len, 1 << log_n);
+
+    for k in 0..len {
+        let rk: u32 = bitreverse(k, log_n);
+        if k < rk {
+            series.swap(rk as usize, k as usize);
+        }
+    }
+
+    let mut m: u32 = 1;
+    for _ in 0..log_n {
+        let w_m = omega.pow(&[(len / (2 * m)) as u64]);
+
+        let mut k = 0;
+        while k < len {
+            let mut w: _ = E::Fr::one();
+            for j in 0..m {
+                let mut t: T = series[(k + j + m) as usize];
+                t *= &w;
+
+                let mut tmp: T = series[(k + j) as usize];
+                tmp -= &t;
+
+                series[(k + j + m) as usize] = tmp;
+                series[(k + j) as usize] += &t;
+
+                w.mul_assign(&w_m);
+            }
+
+            k += 2 * m;
+        }
+
```

Comment:
Remove additional unnecessary white space

Reply:
Agreed, I haven't actually refactored this module except to move it into its own file.

Reply:
Refactored